Make npm install single threaded

sandersn opened the pull request:
Avoids an apparent race condition in npm's lock files when many `npm
install` processes are running at once. Since Azure pipeline machines
typically have 2 cores, and have been running into this problem
regularly, I decided to eliminate parallelism completely.

It would be better, probably, to run npm install before each test, since
it would stagger installs better. I'll look into that next, but this
will unblock the nightly DT test.

weswigham commented:
@sandersn Wanna give this a once over and merge it?

Ah, also, while there's nothing wrong with how this sharding is done, there _is_ a race condition in how DT is cloned:
If you spawn multiple workers and they snap different DT versions, you could get inconsistent results. If the build task is started on a DT trigger (so `--clone` isn't used) this isn't a problem, however if you then use `--localTs` and clone a TS version in your pipeline, if that version isn't set (eg, it's a branch name like `master`), then _that_ can have a race condition. Synchronizing the clones across jobs can be done at the job-orchestrator level, though, eg, by in a non-parallel phase executed before the parallel one, getting the commit SHA of each thing that needs to be cloned and then setting it for the entire pipeline. Supporting this for the `--clone` param may need a little more work, to allow passing in a DT commit sha to clone.

sandersn commented:
Is the purpose of this making DT runs faster for us, or for DT contributors? This only improves it for us.

On the other hand, the median DT contributor wouldn't benefit from sharding, contra the median TS contributor who wants to run DT. Maybe we can leave types-publisher sharding until we start getting close to the 6 hour limit for react. :)

weswigham commented:
For us, mostly. `@typescript-bot run dt` takes an hour and a half right now.

weswigham commented:
(And in a lot of cases it'd be real nice to get that feedback for iteration faster)

sandersn commented on src/index.ts:
```diff
@@ -155,7 +168,7 @@ const exclude = new Set<string>([
 ]);
 
-async function getAllPackages(typesDir: string): Promise<ReadonlyArray<string>> {
+async function getAllPackages(typesDir: string, sharding: {shardId: number, shardCount: number} | undefined): Promise<ReadonlyArray<string>> {
```
can you de-prefix the type to `{ id: number, count: number }` ? It's obvious from useage everywhere what the names refer to.

weswigham replied:
Alright.

sandersn commented on src/index.ts:
```diff
@@ -48,12 +50,23 @@ if (module.parent === null) { // tslint:disable-line no-null-keyword
+            case "--sharded": {
+                i++;
+                assert(i < argv.length);
+                shardId = Number.parseInt(argv[i]);
```
seems like it would be less confusing to `-1` shardId here rather than in the actual filter.

weswigham replied:
The shard is actually assigned a 1-based index by pipelines - i'd rather decrement that for the modular arithmetic operation, so as to not muddy the meaning.

sandersn replied:
No problem; readability is subjective.

weswigham commented:
> Have you tested this manually?

I've run commands for `--sharded 1 2` and `--sharded 2 2` and validated that they started working on different sets of packages. I did not, ofc, wait for all of those to finish validations 😄

weswigham commented:
I also have a [build definition in progress](https://typescript.visualstudio.com/TypeScript/_build?definitionId=23) to make use of this. I'm currently working on preventing the possible race with `--clone`.

weswigham commented:
@sandersn I just added the change to allow `--clone` to optionally take a ref/sha (so `--clone 0376cb`) to this PR (so it can be used to avoid the race condition I mentioned in my first comment). Wanna look over that, too?

weswigham commented:
@sandersn also since this is your repo I can't actually merge this; that's up to you ❤️

sandersn commented on src/index.ts:
```diff
@@ -155,10 +161,28 @@ async function installAllDependencies(
+        return;
+    }
+    else {
+        const cmd = `git clone https://github.com/DefinitelyTyped/DefinitelyTyped.git --depth 1`;
```
what's with all the template literals???

weswigham replied:
Backticks 4 lyfe. But mostly backtick string > double quote string when you're unsure if you're going to use interpolation or quotes in the string or not, and it's kinda strange to have a list of strings but only have one in backticks and the rest not 🤷‍♂️

sandersn commented:
Oops, right. I forgot.

weswigham commented:
Alright. `@typescript-bot run dt faster` should now work. Rather than running DT on one agent, it runs it on 4. This _does_ currently have the potential for clogging up our build queue a bit more (especially during already high-load times of day), though. I also updated the bot to trigger on [regex matches](https://github.com/weswigham/typescript-bot-test-triggerer/blob/master/GithubCommentReader/index.js#L147) rather than simple `.indexOf` (to allow future tasks to read params from the comment) so if you notice any changes in behavior (like it failing to trigger on some input you expected it to trigger for), let me know~

weswigham commented:
Rapid update: Browsing the output of the [trial run](https://typescript.visualstudio.com/TypeScript/_build/results?buildId=26254) I started, it looks like transitive non-DT dependencies of packages not in the current shard aren't installed :( (So, for example, `chai-enzyme` depends on `react` (which is tested in a different shard) which in turn depends on `csstype` (a non-DT dependency, which wasn't installed because we never ran `npm i` in the `react` folder because another shard is running tests on that folder))

@sandersn is there an easy fix for this, or should I just swap the sharded mode of `dtslint-runner` to installing _all_ dependencies, rather than just dependencies for things being tested?

sandersn commented:
Nope, no easy fix, as far as I know. I think you should install all dependencies.

sandersn committed:
Make npm install single threaded

Avoids an apparent race condition in npm's lock files when many `npm
install` processes are running at once. Since Azure pipeline machines
typically have 2 cores, and have been running into this problem
regularly, I decided to eliminate parallelism completely.

It would be better, probably, to run npm install before each test, since
it would stagger installs better. I'll look into that next, but this
will unblock the nightly DT test.

sandersn commented:
This doesn't seem to be failing right now. I'll close it and keep it around for the future.
